### ISSUE TITLE
Possibly missing an error about tuple name mismatch for an implemented interface via base class

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#14841]
```
        [Fact]
        public static void Test1()
        {
            var source = @"
interface I1<T>{}

class Base : I1<(int a, int b)>{}

class Derived : Base, I1<(int c, int d)> {}

interface I2 : I1<(int a, int b)> {}

class C1 : I2, I1<(int c, int d)> {}

class C2 : I2, I1<(int a, int b)> {}
";
            var comp1 = CreateCompilationWithMscorlib(source, references: s_valueTupleRefs);
            comp1.VerifyDiagnostics();
        }
```

Observed:
```
                // (10,7): error CS8140: 'I1<(int c, int d)>' is already listed in the interface list on type 'C1' with different tuple element names, as 'I1<(int a, int b)>'.
                // class C1 : I2, I1<(int c, int d)> {}
                Diagnostic(ErrorCode.ERR_DuplicateInterfaceWithTupleNamesInBaseList, "C1").WithArguments("I1<(int c, int d)>", "I1<(int a, int b)>", "C1").WithLocation(10, 7)
```

Expected: A similar error about class Derived because base implements an interface with different tuple names.

CC @VSadov, @jcouv, @agocke

**Comment by jcouv** [comment]
@AlekseyTs, no error is reported in the related scenario of type unification through a base class. I'll stop by to discuss the implications for this issue.

```C#
        [Fact]
        public static void TypeUnificationThroughBaseClass()
        {
            var source = @"
interface I1<T>{}
class Base : I1<int>{}
class Derived<T> : Base, I1<T> {}
";
            var comp1 = CreateCompilationWithMscorlib(source);
            comp1.VerifyDiagnostics();
        }
```

**Comment by jcouv** [comment]
More details on this change, for the record:

We are making a change to C# and VB to allow re-implementating interfaces with different tuple names, when it is on another inheritance level:
```C#
public class Base : I0<(int a, int b)> { }
public class Derived : Base, I0<(int notA, int not B)> { } // re-implementation ok
public class Derived2<T> Base, I0<T> { } // possible re-implementation ok
```

Both `Derived` and `Derived2<(int notA, int notB)>` only implement one interface: `I0<(int notA, int notB)>`. 
This is observable if the interface is `IEnumerable`, as you will see tuple names “notA” and “notB” when iterating on those derived types.

The motivation for this is that the compiler doesn’t do such checks (interface duplication or unification) across inheritance levels with non-tuple types, in accordance to section §13.4.2 of the spec. 
So it would be odd and costly to implement such checks just for tuple names.

For instance:
```C#
public class Base : I0<int> { }
public class Derived : Base, I0<int> { } // re-implementation ok
public class Derived2<T> Base, I0<T> { } // possible re-implementation ok
```


Note 1: 
We are retaining the error when the same interface is implemented multiple times (ignoring tuple names) 

1)	On a given inheritance level
```C#
public interface I0<T> { }
public class C1 : I0<(int a, int b)>, I0<(int notA, int notB)> { } // error
public class C2 : I0<(int a, int b)>, I0<(int a, int b)> { } // error
public class C4<T, U> : I0<(int a, int b)>, I0<(T a, U b)> { } // error
```
2)	Via interface implementations
```C#
public interface I1 : I0<(int a, int b)> { }
public interface I2 : I0<(int notA, int notB)> { }
public interface I3 : I0<(int a, int b)> { }

public class C : I1, I2 { } // error
public class D : I1, I3 { } // ok
```

Note 2: 
When it comes to Derived trying to re-implement the interface, it is still limited: the re-implementation must use “notA” and “notB”. 

This means that a public method inherited from Base doesn’t count, as in this example:
```C#
class Base : I<(int a, int b)>
{
    public (int a, int b) M() { return (1, 2); }
}
class Derived : Base, I<(int notA, int notB)>
{
    // error
}
```
Note 3:
This also means we can get slightly surprising behaviors like (tuple names from Base, but implementation from Derived):
```C#
static void Main()
    {
        Base b = new Derived();
        var x = Test(b); // tuple names from Base
        System.Console.Write(x.a); // prints value from Derived.M
    }
static T Test<T>(I<T> t) { return t.M(); }
```

**Comment by jcouv** [comment]
Re-opening to track further investigation based on LDM feedback.

**Comment by jcouv** [comment]
After further exploration and discussion, we decided that the implemented behavior (most derived set of names wins) is ok. Closing again.
